Blocking repeated ngrams during beam search

danieldeutsch opened the pull request:
<!-- To ensure we can review your pull request promptly please complete this template entirely. -->

<!-- Please reference the issue number here. You can replace "Fixes" with "Closes" if it makes more sense. -->
Closes #5205.

Changes proposed in this pull request:
<!-- Please list all changes/additions here. -->
- Adds a new `Constraint` abstract class for enforcing constraints during beam search
- Adds a `NGramBlockingConstraint` class to prevent decoding repeated ngrams.

This is a work in progress and maybe close to finishing, so I wanted to get feedback. @epwalsh, any thoughts on
1. The `Constraint` interface? I limited the methods to just those which I needed to implement the ngram blocking
2. How to more efficiently implement the ngram blocking? I'm not sure there's a way to get around maintaining which ngrams have appeared before in a dictionary.
3. How to test the end-to-end beam search with the ngram blocking constraint? I've tested the individual methods and some toy examples on my own. To really test it, I would need to come up with a transition matrix which has repeated ngrams by default and then block them. I couldn't think of a simple one which wouldn't require a lot of manual effort to ensure the output is correct.

epwalsh reviewed (commented):
This looks pretty good so far. I do think having an end-to-end test is important. Is it feasible to at least make a test that assumes `ngram_size` is 1?

One comment I have about the API for the `Constraint` class is that having to deal with `backpointers` in `update_state()` is not ideal. Can we automatically handle reordering the `state` list outside of `update_state()`, either in the `BeamSearch` class or the `Constraint` base class?

danieldeutsch commented:
In this version:
- I incorporated your comments on the code
- I moved the logic to copy the parent's state into the `Constraint` class
- I added end-to-end unit tests for using the n-gram blocking in `BeamSearch` with a search that I manually traced. See `test_take_repeated_ngram_step`

I removed "WIP" from the PR title because I think this version is complete to me, pending comments.

epwalsh commented on CHANGELOG.md:
```diff
@@ -33,6 +33,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added a `min_steps` parameter to `BeamSearch` to set a minimum length for the predicted sequences.
 - Added the `FinalSequenceScorer` abstraction to calculate the final scores of the generated sequences in `BeamSearch`. 
 - Added `shuffle` argument to `BucketBatchSampler` which allows for disabling shuffling.
+- Added a `Constrant` abstract class to `BeamSearch`, which allows for incorporating constraints on the predictions found by `BeamSearch`.
```
```suggestion
- Added a `Constraint` abstract class to `BeamSearch`, which allows for incorporating constraints on the predictions found by `BeamSearch`.
```

epwalsh commented on allennlp/nn/beam_search.py:
```diff
@@ -524,6 +527,176 @@ def score(
+class Constraint(Registrable):
+    """
+    An abstract class that can be used to enforce constraints on the output predictions
+    by manipulate the class log probabilities during beam search.
```
```suggestion
    by manipulating the class log probabilities during beam search.
```

epwalsh commented on allennlp/nn/beam_search.py:
```diff
@@ -524,6 +527,176 @@ def score(
+
+    The `apply()` method should manipulate the `class_log_probabilities` in place to enforce the constraint
+    for this step of beam search. For instance, it may prevent a specific class from being selected by setting
+    the corresponding log probability to `-inf` (by using `min_value_of_dtype(class_log_probabilities.dtype)`).
```
```suggestion
    the corresponding log probability to a negligible value by using `min_value_of_dtype(class_log_probabilities.dtype)`, which is essentially equivalent to `-inf`.
```

epwalsh commented on allennlp/nn/beam_search.py:
```diff
@@ -792,6 +984,13 @@ def _search(
             # shape: (batch_size * beam_size, num_classes)
             class_log_probabilities, state = step(last_predictions, state, timestep + 1)
 
+            for constraint, constraint_state in zip(self.constraints, constraint_states):
+                # shape: (batch_size, beam_size, num_classes)
+                reshaped_class_log_probabilities = class_log_probabilities.view(
+                    batch_size, self.beam_size, -1
+                )
```
Any reason not to put the reshape outside of the loop?

```suggestion
            # shape: (batch_size, beam_size, num_classes)
            reshaped_class_log_probabilities = class_log_probabilities.view(
                batch_size, self.beam_size, -1
            )
            for constraint, constraint_state in zip(self.constraints, constraint_states):
```

epwalsh commented on allennlp/nn/beam_search.py:
```diff
@@ -524,6 +527,176 @@ def score(
+    - `class_log_probabilities`, a tensor of shape `(batch_size, beam_size, num_classes)` that contains the
+    log probabilities for the classes during search. The first time `apply()` is called, `beam_size = 1`.
+
+    The `apply()` method should manipulate the `class_log_probabilities` in place to enforce the constraint
```
I think it's more natural for the `apply()` method to return new `class_log_probabilities`. It's a little easier to reason about.

epwalsh reviewed (changes requested):
Looks great @danieldeutsch! Just a few more minor comments.

epwalsh reviewed (approved):
LGTM, thanks @danieldeutsch!

danieldeutsch commented:
Thanks @epwalsh for finishing it. I forgot about it over the holiday

JohnGiorgi commented on allennlp/nn/beam_search.py:
```diff
@@ -524,6 +527,162 @@ def score(
+                    backpointer = last_backpointer[i, j].item()
+                batch_state.append(copy.deepcopy(state[i][backpointer]))
+            new_state.append(batch_state)
+        return new_state
```
@danieldeutsch @epwalsh  Sorry, I know this PR is closed but I have a question that's easiest to ask here because of the existing context.

It seems that `last_backpointer`, which is passed in by `update` will always be `None` (its never provided to `update` in `BeamSearch`). That would mean that `backpointer` will always by 0, and then then it will always be `state[i][0]` being copied, regardless of the actual timestep. Isn't this a problem? The tests don't catch this because they provide `backpointer` to `update` manually (see my other comment).

danieldeutsch replied:
Yes, it looks like you are right to me. I think the fix should be to pass `backpointer` here, right?

https://github.com/allenai/allennlp/blob/5b1da908238794efb318da3327271cc945c6ba46/allennlp/nn/beam_search.py#L1095-L1098

JohnGiorgi commented on tests/nn/beam_search_test.py:
```diff
@@ -596,3 +630,202 @@ def test_length_normalized_sequence_log_prob_scorer(self):
+            ],
+        ]
+        predictions = torch.LongTensor([[5, 6], [0, 3]])
+        backpointers = torch.LongTensor([[1, 1], [0, 1]])
```
`backpointers` provided here to `update`, but not in `BeamSearch` (see my other comment).

danieldeutsch replied:
I would need to retrace the beam search that I used to write this test, but my guess is that this is not caught because the backpointer is always 0
https://github.com/allenai/allennlp/blob/5b1da908238794efb318da3327271cc945c6ba46/tests/nn/beam_search_test.py#L774
